src/gt4py/next/common.py

```diff
@@ -117,10 +129,102 @@ def __getitem__(self, index: int | slice) -> int | UnitRange:
             else:
                 raise IndexError("UnitRange index out of range")
 
+    def __and__(self, other: UnitRange) -> UnitRange:
+        start = max(self.start, other.start)
+        stop = min(self.stop, other.stop)
+        return UnitRange(start, stop)
+
 
 DomainT: TypeAlias = tuple[tuple[Dimension, UnitRange], ...]
 
 
+from dataclasses import dataclass
+from collections.abc import Sequence
+
+
+@dataclass(frozen=True)
+class Domain(Sequence):
+    dims: list[Dimension]
+    ranges: list[UnitRange]
```

Comment:
I used `list` as otherwise mypy complains since `promote_dims` expects a list of Dimensions so I just adapted to that interface to avoid having to change type hints everywhere.

```
src/gt4py/next/common.py:183:39: error: Argument 1 to "promote_dims" has incompatible type "Tuple[Dimension]"; expected "List[Dimension]"  [arg-type]
```

Reply:
Then let's change the interface of promote dims to

```
promote_dims(*dims_list: Sequence[Dimension]) -> list[Dimension]
```